[XBDM] Implement DmQueryTitleMemoryStatistics

Doliman100 opened the pull request:
I didn't find a way to access the structure and function definitions from xboxkrnl_memory.cc from xbdm_misc.cc file. As a workaround I copied them into into xbdm_misc.cc. If someone know how to get rid of double definition of these structures and function declaration, please let me know how to fix this.

Doliman100 committed:
[XBDM] Implement DmQueryTitleMemoryStatistics

Gliniak commented:
You can just move them from .cc file to xboxkrnl_memory.h that way you will prevent structures duplication

Doliman100 commented:
Thank you, I moved the structure definitions to the header and included it into xbdm_misc.cc. But I'm not sure what to do with the `MmQueryStatistics_entry` function declaration. Should I also move it to the header, or is there another way to call exported kernel functions? I couldn't find any other header declaring any *_entry functions to use as an example.

Gliniak commented:
Look how: xeMmAllocatePhysicalMemoryEx and MmAllocatePhysicalMemoryEx_entry works.

Gliniak commented on src/xenia/kernel/xbdm/xbdm_misc.cc:
```diff
@@ -465,6 +477,51 @@ dword_result_t DmSetDumpMode_entry(dword_t dump_mode) {
+
+dword_result_t DmQueryTitleMemoryStatistics_entry(
+    pointer_t<X_DM_QUERY_MEMORY_STATISTICS_RESULT> stats_ptr) {
+  if (stats_ptr->size != 48) return X_E_INVALIDARG;
```
replace magic number with sizeof(X_DM_QUERY_MEMORY_STATISTICS_RESULT)

Gliniak commented on src/xenia/kernel/xbdm/xbdm_misc.cc:
```diff
@@ -465,6 +477,51 @@ dword_result_t DmSetDumpMode_entry(dword_t dump_mode) {
+  if (stats_ptr->size != 48) return X_E_INVALIDARG;
+
+  xe::kernel::xboxkrnl::X_MM_QUERY_STATISTICS_RESULT stats;
+  stats.size = 104;
```
Same thing. You can use sizeof instead of hardcoded value

Gliniak commented on src/xenia/kernel/xbdm/xbdm_misc.cc:
```diff
@@ -465,6 +477,51 @@ dword_result_t DmSetDumpMode_entry(dword_t dump_mode) {
+  xe::kernel::xboxkrnl::X_MM_QUERY_STATISTICS_RESULT stats;
+  stats.size = 104;
+  dword_result_t result = xe::kernel::xboxkrnl::MmQueryStatistics_entry(&stats);
+  if (result < 0) return result | 0x10000000;  // HRESULT_FROM_NT(result)
```
This mask is kinda strange. If it is some base status mask then it should be defined somewhere

Doliman100 replied:
Thank you for your notes. I've added X_HRESULT_FROM_NT macro and X_FACILITY_NT_BIT constant.

Doliman100 commented:
I renamed the function to `xeMmQueryStatistics` and added a declaration to the header. Not sure if it allowed to use `dword_result_t` and `pointer_t` types there, but since this function calls `pointer_t::Zero()` method, I kept them as is. Should I replace it with `X_MM_QUERY_STATISTICS_RESULT *stats_ptr` and `std::memset(stats_ptr, 0, sizeof(X_MM_QUERY_STATISTICS_RESULT))`?

Gliniak commented on src/xenia/kernel/xbdm/xbdm_misc.cc:
```diff
@@ -465,6 +466,52 @@ dword_result_t DmSetDumpMode_entry(dword_t dump_mode) {
+
+dword_result_t DmQueryTitleMemoryStatistics_entry(
+    pointer_t<X_DM_QUERY_MEMORY_STATISTICS_RESULT> stats_ptr) {
+  if (stats_ptr->size != sizeof(X_DM_QUERY_MEMORY_STATISTICS_RESULT))
```
Project style requires such cases to be wrapped around brackets.
if ([something]) { [statement] }

Doliman100 replied:
Good, I added brackets.

Gliniak commented on src/xenia/kernel/xbdm/xbdm_misc.cc:
```diff
@@ -465,6 +466,52 @@ dword_result_t DmSetDumpMode_entry(dword_t dump_mode) {
+  xe::kernel::xboxkrnl::X_MM_QUERY_STATISTICS_RESULT stats;
+  stats.size = sizeof(xe::kernel::xboxkrnl::X_MM_QUERY_STATISTICS_RESULT);
+  dword_result_t result = xe::kernel::xboxkrnl::xeMmQueryStatistics(&stats);
+  if (result < 0) return X_HRESULT_FROM_NT(result);
```
Same thing, missing brackets {}

Gliniak commented:
> I renamed the function to `xeMmQueryStatistics` and added a declaration to the header. Not sure if it allowed to use `dword_result_t` and `pointer_t` types there, but since this function calls `pointer_t::Zero()` method, I kept them as is. Should I replace it with `X_MM_QUERY_STATISTICS_RESULT *stats_ptr` and `std::memset(stats_ptr, 0, sizeof(X_MM_QUERY_STATISTICS_RESULT))`?

No, you should not. We're still working kinda in the guest domain, so we should use wrappers for guest types where it is possible

Gliniak commented:
Last thing. Please squash these commits into one

Doliman100 commented:
Done.
